We should prevent RTCPeerConnection destructor to be blocked on webrtc::PeerConnection destructor

youennf opened the pull request:
#### f20ef707eabe48e88479c0e4167332b94876890a
<pre>
We should prevent RTCPeerConnection destructor to be blocked on webrtc::PeerConnection destructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=297590">https://bugs.webkit.org/show_bug.cgi?id=297590</a>
<a href="https://rdar.apple.com/157701507">rdar://157701507</a>

Reviewed by Jean-Yves Avenard.

PeerConnection destructor may be costly so we make sure to not block main thread by hopping m_backend reference to signalling thread before resetting it.

* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::stop):

Canonical link: <a href="https://commits.webkit.org/299000@main">https://commits.webkit.org/299000@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f5df52a8c46e0e6a593c6cc42e5542ff6e8d777

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117452 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37122 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123552 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69443 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bac79892-0d89-4ce0-b664-c207b697c173) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119330 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45709 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89133 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4af8049e-da04-4389-b104-10ce0ba46a67) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120404 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30110 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105314 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69642 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c7ce7e74-13a4-4930-adf0-2d8e3387af96) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29166 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23428 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67226 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99515 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23610 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126672 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44349 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97799 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44707 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101548 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97593 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24842 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42965 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20901 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44222 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49881 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43679 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47023 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45374 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
```diff
@@ -92,10 +92,16 @@ LibWebRTCMediaEndpoint::LibWebRTCMediaEndpoint(LibWebRTCPeerConnectionBackend& p
+LibWebRTCMediaEndpoint::~LibWebRTCMediaEndpoint()
+{
+    // We move m_backend to the signalling thread in case we own the last reference, as deallocating m_backend could block on the signalling thread.
+    LibWebRTCProvider::callOnWebRTCSignalingThread([backend = std::exchange(*const_cast<RefPtr<webrtc::PeerConnectionInterface>*>(&m_backend), { })]() mutable { });
```
```suggestion
   LibWebRTCProvider::callOnWebRTCSignalingThread([backend = std::exchange(*const_cast<RefPtr<webrtc::PeerConnectionInterface>*>(&m_backend), { })] { });
```

youennf replied:
I rebased over main instead of the first PR, so no longer need for the const_cast.

youennf committed:
We should prevent RTCPeerConnection destructor to be blocked on webrtc::PeerConnection destructor
https://bugs.webkit.org/show_bug.cgi?id=297590
rdar://157701507

Reviewed by Jean-Yves Avenard.

PeerConnection destructor may be costly so we make sure to not block main thread by hopping m_backend reference to signalling thread before resetting it.

* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::stop):

Canonical link: https://commits.webkit.org/299000@main
